src/main/kotlin/com/sourcegraph/cody/history/HistoryTree.kt

```diff
@@ -1,84 +1,125 @@
+import java.awt.event.ActionEvent
+import java.awt.event.KeyEvent
+import javax.swing.*
 import javax.swing.tree.DefaultMutableTreeNode
 import javax.swing.tree.DefaultTreeModel
+import javax.swing.tree.TreeSelectionModel
 
-class HistoryTree(onSelect: (ChatState) -> Unit, onDelete: (ChatState) -> Unit) : Tree() {
+class HistoryTree(
+    private val onSelect: (ChatState) -> Unit,
+    private val onDelete: (ChatState) -> Unit
+) : SimpleToolWindowPanel(true, true) {
 
-  private val popup = HistoryPopupMenu(onSelect, onDelete, getSelected = { getLeafOrNull()!!.chat })
+  // todo commit msg: Deleting chats no longer resets the tree view to its initial state
+  // todo commit msg: The background color is now the same as other toolbars in IDE.
+  // todo commit msg: You can now search by titles if you are focused on the panel (just start
+  // typing).
+  // todo commit msg: Added separator to popup menu because it was easy to miss-click delete instead
+  // of select.
+  // todo commit msg: The popup menu and select/delete keyboard actions now registered as IDE
+  // actions (instead of Swing events).
+
+  private val model = DefaultTreeModel(buildTree())
+  private val tree =
+      SimpleTree(model).apply {
+        isRootVisible = false
+        cellRenderer = HistoryTreeNodeRenderer()
+        selectionModel.selectionMode = TreeSelectionModel.SINGLE_TREE_SELECTION
+        inputMap.put(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), ENTER_MAP_KEY)
+        inputMap.put(KeyStroke.getKeyStroke(KeyEvent.VK_DELETE, 0), DELETE_MAP_KEY)
+        actionMap.put(ENTER_MAP_KEY, ActionWrapper(::selectSelected))
+        actionMap.put(DELETE_MAP_KEY, ActionWrapper(::deleteSelected))
+      }
 
   init {
-    isRootVisible = false
-    addMouseListener(DoubleLeftClickListener { invokeOnlyOnLeaf { onSelect(it) } })
-    addMouseListener(
-        RightClickListener { mouse -> invokeOnlyOnLeaf { popup.show(this, mouse.x, mouse.y) } })
-    addKeyListener(EnterListener { invokeOnlyOnLeaf { onSelect(it) } })
-    addKeyListener(DeleteListener { invokeOnlyOnLeaf { onDelete(it) } })
+    val group = DefaultActionGroup()
+    group.add(LeafPopupAction(tree, "Select Chat", null, ::selectSelected))
+    group.addSeparator()
+    group.add(LeafPopupAction(tree, "Remove Chat", AllIcons.Actions.GC, ::deleteSelected))
+    PopupHandler.installPopupMenu(tree, group, "ChatActionsPopup")
+    EditSourceOnDoubleClickHandler.install(tree, ::selectSelected)
+    setContent(ScrollPaneFactory.createScrollPane(tree))
+  }
+
+  fun refreshTree() {
+    // todo: rename to "notifyAdded" and use in such context when model is changed
+    // todo: this method must be executed when a new chat is added
+    // todo: use model.insertNodeInto(child, parent)
   }
 
-  override fun getToolTipText(event: MouseEvent): String {
-    val leaf = getPathForLocation(event.x, event.y)?.lastPathComponent as? ChatLeafNode
+  private fun selectSelected() {
+    val leaf = tree.selectedLeafOrNull()
```

Comment:
Whatever you like more, but as new-to-kotlin person I found those one-liners quite nice:
`tree.selectedLeafOrNull()?.let { onSelect(leaf.chat) }`

Reply:
Agreed. Fixed!

BTW funny thing, this code works:

```kotlin
private fun selectSelected() {
  tree.selectedLeafOrNull()?.let { onSelect(it.chat) }
}
```

But this code can't be compiled due to _type checking recursive problem_:

```kotlin
private fun selectSelected() = tree.selectedLeafOrNull()?.let { onSelect(it.chat) }
```